Add bag data structure

jacob-hughes opened the pull request:
A bag is an vector-like data structure using aligned block which is optimised for constant time insertion and removal.

Removal is O(1) because a removed item will leave a hole in the bag rather than shifting remaining items down like in a bag. To reduce fragmentation which can occur with deallocation churn, the bag maintains a freelist which is threaded through the holes left in the bag. Insertion of new items will first query the freelist to see if a hole can be reused, otherwise items are bump allocated.

This data structure is useful for tracking objects in regions -- such as one which uses reference counting -- where efficient alloc/dealloc is necessary.

jacob-hughes commented:
This PR is a WIP because I'm still working through a bug with the `INDEX_MASK`. Once this is fixed I'll push additional tests too. But I think it's useful @mjp41 for you to comment on the new union based approach for elements inside the bag.

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,255 @@
+      if (next_free != nullptr)
+      {
+        MaybeElem* prev = next_free->hole_ptr;
+        next_free->item = item;
```
Can you assert here that the bit is set.

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,246 @@
+
+      inline iterator end()
+      {
+        return {bag, nullptr};
```
```suggestion
        return {bag, bag->null_index};
```

jacob-hughes replied:
Why do we want to return `null_index` here? I thought the iterator advancement will return an `Elem*`, and when no more elements exist will return `nullptr`. Isn't exposing `null_index` (which is a `MaybeElem*`) an implementation detail?

mjp41 replied:
If you do this, you will end up with less.  Code, the bag implementation uses `null_index` to signify when it has reached the end, but your implementation of iterator uses a different constant to signify the end.

This is not being exposed, it is the value of a private field.

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,246 @@
+          p--;
+          if (p->hole_ptr == bag->null_index)
+          {
+            ptr = nullptr;
+            return *this;
```
```suggestion
            break;
```

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,246 @@
+            p = bag->get_block(p)->prev.hole_ptr;
+          }
+        } while ((uintptr_t)p->hole_ptr & EMPTY_MASK);
+
```
```suggestion
        assert(
          (p == bag.null_index) || 
          (((uintptr_t)p->hole_ptr & EMPTY_MASK) == 0));
```

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,246 @@
+        do
+        {
+          p--;
+          if (p->hole_ptr == bag->null_index)
```
```suggestion
          if (p->hole_ptr == Bag::null_index)
```

mjp41 commented on src/rt/test/func/bag/bag.cc:
```diff
@@ -0,0 +1,84 @@
+  // Check that the freelist is used to fill existing holes before bump
+  // allocation.
+  auto idx1 = bag.insert({nullptr, 123}, alloc);
+  assert((uintptr_t)rm1 == (uintptr_t)idx1);
```
This is very strict test that imposes a stack like ordering of the free list.  Is there any easy way to make it only impose the constraints we really want.

How about effectively saying the sets are equal? `{rm1, rm2, rm3} == {idx1,idx2,idx3}`

jacob-hughes replied:
What about something like this ? 57cbf756da87743cd0237dfb6989bff8a9462279

mjp41 commented on src/rt/test/func/bag/bag.cc:
```diff
@@ -0,0 +1,84 @@
+
+  auto idx3 = bag.insert({nullptr, 789}, alloc);
+  assert((uintptr_t)rm3 == (uintptr_t)idx3);
+
```
Can you allocate one more that won't be using free list at this point to get the coverage for changing mode.

jacob-hughes replied:
Added as part of 57cbf756da87743cd0237dfb6989bff8a9462279

mjp41 commented on src/rt/test/func/bag/bag.cc:
```diff
@@ -0,0 +1,87 @@
+    UNUSED(item);
+    count++;
+  }
+  assert(count == NUM_OBJECTS - 3);
```
Don't use assert in tests.  This wants to be checked on release builds too.   We have a macro that does `check`s:
```suggestion
  check(count == NUM_OBJECTS - 3);
```

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,250 @@
+    public:
+      iterator(Bag<T, U, Alloc>* bag) : bag(bag)
+      {
+        // If the bag is empty then there is nothing to iterate over.
+        if (bag->empty())
+        {
+          ptr = nullptr;
+          return;
+        }
+
```
If you used `null_index` as end, then this code would go away.

mjp41 commented on src/rt/test/func/bag/bag.cc:
```diff
@@ -0,0 +1,95 @@
+void test()
+{
+  auto& alloc = ThreadAlloc::get();
+
```
Perhaps add
```C++
using B = Bag<uintptr_t, uintptr_t, Alloc>;
using E = B::Elem*;
```
and then replace the uses below.

mjp41 commented on src/rt/test/func/bag/bag.cc:
```diff
@@ -0,0 +1,95 @@
+
+  check(count == NUM_OBJECTS);
+
+  std::set<uintptr_t> removed = {
```
If you add the shorthands above, then this can be:
```suggestion
  std::set<E> removed = {
```
and then you can drop all the casts to `uintptr_t`.  Or is this because you need a less defined on the type?  Does `unordered_set` help here?  You aren't traversing in an order anyway?

jacob-hughes replied:
I need a less defined type here as I just want to compare the pointer addresses. I've changed it to an `unordered_set` though: 85ae978cb22f4acb0bf1ffd956f62d9e95b4282f

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,250 @@
+
+    private:
+      Bag<T, U, Alloc>* bag;
+      Elem* ptr;
```
I guess given my suggestion this would have to be
```suggestion
      MaybeElem* ptr;
```
```suggestion
      Elem* ptr;
```

mjp41 commented on src/rt/ds/bag.h:
```diff
@@ -0,0 +1,250 @@
+
+      inline Elem* operator*() const
+      {
+        return ptr;
```
This would have to change to
```suggestion
        return ptr->item;
```
```suggestion
        return ptr;
```

jacob-hughes commented:
Addressed comments here 85ae978cb22f4acb0bf1ffd956f62d9e95b4282f and refactored the pointer while loop into a `next_non_empty` function. I've also removed the `peek` function because I think it's misleading as it implies stack like behaviour on the Bag: we make no such ordering guarantees so the last inserted item may not be what `peek` returns.

jacob-hughes commented:
I've reworked the `next_non_empty` method to use a `step` method instead now, and removed the misleading `empty` method.
